src/TelemetryChannels/ServerTelemetryChannel/Shared/Implementation/SamplingPercentageEstimatorTelemetryProcessor.cs

```diff
@@ -0,0 +1,244 @@
+        public void Process(ITelemetry item)
+        {
+            // increment post-samplin telemetry item counter
+            this.itemCount.Increment();
+
+            // continue processing telemetry item with the next telemetry processor
+            this.next.Process(item);
+        }
+
+        /// <summary>
+        /// Disposes the object.
+        /// </summary>
+        public void Dispose()
+        {
+            if (this.evaluationTimer != null)
+            {
+                this.evaluationTimer.Dispose();
+                this.evaluationTimer = null;
+            }
+        }
+
+        /// <summary>
+        /// Checks to see if exponential moving average has changed.
+        /// </summary>
+        /// <param name="running">Currently running value of moving average.</param>
+        /// <param name="current">Value set in the algorithm parameters.</param>
+        /// <returns>True if moving average value changed.</returns>
+        private static bool MovingAverageCoefficientChanged(double running, double current)
+        {
+            const double Precision = 1E-12;
+
+            return (running < current - Precision) || (running > current + Precision);
+        }
+
+        /// <summary>
+        /// Callback for sampling percentage evaluation timer.
+        /// </summary>
+        /// <param name="state">Timer state.</param>
+        private void EstimateSamplingPercentage(object state)
+        {
+            // get observed after-sampling eps
+            double observedEps = this.itemCount.StartNewInterval() / this.evaluationInterval.TotalSeconds;
+
+            // we see events post sampling, so get pre-sampling eps
+            double beforeSamplingEps = observedEps * this.currenSamplingRate;
+
+            // caclulate suggested sampling rate
+            int suggestedSamplingRate = (int)Math.Ceiling(beforeSamplingEps / this.settings.EffectiveMaxTelemetryItemsPerSecond);
+
+            // adjust suggested rate so that it fits between min and max configured
+            if (suggestedSamplingRate > this.settings.EffectiveMaxSamplingRate)
+            {
+                suggestedSamplingRate = this.settings.EffectiveMaxSamplingRate;
+            }
+
+            if (suggestedSamplingRate < this.settings.EffectiveMinSamplingRate)
+            {
+                suggestedSamplingRate = this.settings.EffectiveMinSamplingRate;
+            }
+
+            // see if evaluation interval was changed and apply change
+            if (this.evaluationInterval != this.settings.EffectiveEvaluationInterval)
+            {
+                this.evaluationInterval = this.settings.EffectiveEvaluationInterval;
+                this.evaluationTimer.Change(this.evaluationInterval, this.evaluationInterval);
+            }
+
+            // check to see if sampling rate needs changes
+            bool samplingPercentageChangeNeeded = suggestedSamplingRate != this.currenSamplingRate;
```

Comment:
Sampling rates are INTs expressing sampling percentage in form of 1/N (take 1 of every N items). So, they can be compared w/o precision...